From 30e4e21b41fba230b0455ae30fabcd12131703ac Mon Sep 17 00:00:00 2001
From: "U. Artie Eoff" <ullysses.a.eoff@intel.com>
Date: Wed, 6 Jul 2022 09:08:17 -0400
Subject: [PATCH 11/11] kmssink: implement set_window_handle interface

Implement the interface for GstVideoOverlay's set_window_handle.

This allows an application to provide their own opened DRM device
fd handle to kmssink.  For example, an application can lease
multiple fd's from a DRM master to display on different CRTC
outputs at the same time with multiple kmssink instances.
---
 .../gst-plugins-bad/sys/kms/gstkmssink.c      | 48 ++++++++++++++++---
 .../gst-plugins-bad/sys/kms/gstkmssink.h      |  3 ++
 2 files changed, 45 insertions(+), 6 deletions(-)

diff --git a/subprojects/gst-plugins-bad/sys/kms/gstkmssink.c b/subprojects/gst-plugins-bad/sys/kms/gstkmssink.c
index 788cefc06970..fd5900647abf 100644
--- a/subprojects/gst-plugins-bad/sys/kms/gstkmssink.c
+++ b/subprojects/gst-plugins-bad/sys/kms/gstkmssink.c
@@ -165,9 +165,28 @@ gst_kms_sink_expose (GstVideoOverlay * overlay)
   gst_kms_sink_show_frame (GST_VIDEO_SINK (self), NULL);
 }
 
+static void
+gst_kms_sink_set_window_handle (GstVideoOverlay * overlay, guintptr id)
+{
+  GstKMSSink *self = GST_KMS_SINK (overlay);
+
+  g_return_if_fail (self->fd < 0);
+  g_return_if_fail (!self->is_running);
+
+  /* If the id is 0, we use an internal device fd */
+  if (id == 0) {
+    self->fd = -1;
+    self->is_internal_fd = TRUE;
+  } else {
+    self->fd = (gint) id;
+    self->is_internal_fd = FALSE;
+  }
+}
+
 static void
 gst_kms_sink_video_overlay_init (GstVideoOverlayInterface * iface)
 {
+  iface->set_window_handle = gst_kms_sink_set_window_handle;
   iface->expose = gst_kms_sink_expose;
   iface->set_render_rectangle = gst_kms_sink_set_render_rectangle;
 }
@@ -729,10 +748,18 @@ gst_kms_sink_start (GstBaseSink * bsink)
   pres = NULL;
   plane = NULL;
 
-  if (self->devname || self->bus_id)
-    self->fd = drmOpen (self->devname, self->bus_id);
-  else
-    self->fd = kms_open (&self->devname);
+  /* notify application to set device fd handle now */
+  if (self->fd < 0)
+    gst_video_overlay_prepare_window_handle (GST_VIDEO_OVERLAY (self));
+
+  /* open our own internal device fd if application did not */
+  if (self->is_internal_fd) {
+    if (self->devname || self->bus_id)
+      self->fd = drmOpen (self->devname, self->bus_id);
+    else
+      self->fd = kms_open (&self->devname);
+  }
+
   if (self->fd < 0)
     goto open_failed;
 
@@ -823,6 +850,7 @@ retry_find_plane:
   gst_kms_sink_update_connector_properties (self);
   gst_kms_sink_update_plane_properties (self);
 
+  self->is_running = TRUE;
   ret = TRUE;
 
 bail:
@@ -838,8 +866,11 @@ bail:
     drmModeFreeResources (res);
 
   if (!ret && self->fd >= 0) {
-    drmClose (self->fd);
+    if (self->is_internal_fd) {
+      drmClose (self->fd);
+    }
     self->fd = -1;
+    self->is_running = FALSE;
   }
 
   return ret;
@@ -945,10 +976,13 @@ gst_kms_sink_stop (GstBaseSink * bsink)
   }
 
   if (self->fd >= 0) {
-    drmClose (self->fd);
+    if (self->is_internal_fd)
+      drmClose (self->fd);
     self->fd = -1;
   }
 
+  self->is_running = FALSE;
+
   GST_OBJECT_LOCK (bsink);
   self->hdisplay = 0;
   self->vdisplay = 0;
@@ -1906,6 +1940,8 @@ static void
 gst_kms_sink_init (GstKMSSink * sink)
 {
   sink->fd = -1;
+  sink->is_internal_fd = TRUE;
+  sink->is_running = FALSE;
   sink->conn_id = -1;
   sink->plane_id = -1;
   sink->can_scale = TRUE;
diff --git a/subprojects/gst-plugins-bad/sys/kms/gstkmssink.h b/subprojects/gst-plugins-bad/sys/kms/gstkmssink.h
index 428c057e7764..a64be2e75d23 100644
--- a/subprojects/gst-plugins-bad/sys/kms/gstkmssink.h
+++ b/subprojects/gst-plugins-bad/sys/kms/gstkmssink.h
@@ -93,6 +93,9 @@ struct _GstKMSSink {
   /* reconfigure info if driver doesn't scale */
   GstVideoRectangle pending_rect;
   gboolean reconfigure;
+
+  gboolean is_internal_fd;
+  gboolean is_running;
 };
 
 struct _GstKMSSinkClass {
-- 
2.34.3

